### PR TITLE
[Unification] Move masks out of attention modules and return attn probs

### PR DESCRIPTION
**Pull request by RdoubleA**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #172
* #170
* #169
* #168
* __->__ #162
* #152
* #151

## Summary
Small diffs building up to a unified `SelfAttention` class used across all models by generalizing one piece at a time.

This diff moves decoding masks out of attention modules into a user specified argument. We can then simplify the `FullAttention` API. This diff also updates attention modules to return the attention probabilities as well. The [original PR](https://github.com/facebookresearch/multimodal/pull/167) got squashed with this while running rebase.

## Test plan
`pytest test -vv`
`pytest examples -vv`

Differential Revision: [D37938513](https://our.internmc.facebook.com/intern/diff/D37938513)

### CODE REVIEW

File: torchmultimodal/modules/layers/attention.py [hunk 1]
```diff
@@ -194,26 +194,18 @@ class FullAttention(nn.Module):
```
Comment (langong347) [inline torchmultimodal/modules/layers/attention.py:194]: Add docstring for `attention_mask` and `head_mask` with dimensions.
The same for `MultiHeadAttention`.

**Review by langong347 — COMMENTED** [review]
In a follow-up PR: add unit tests for `scaled_dot_product_attention` with `attention_mask` and `head_mask`.

**Review by langong347 — APPROVED** [review]

File: torchmultimodal/modules/layers/attention.py [hunk 2]
```diff
@@ -13,65 +13,111 @@
...
+        attn_dropout (float): probability of dropout after softmax. Default is ``0.0``.
 
-    Follows implementation by VideoGPT:
-    https://github.com/wilson1yan/VideoGPT/blob/master/videogpt/vqvae.py
+    Args:
+        q, k, v (Tensor): a [b, h, d1, ..., dn, c] tensor where h is the number of attention heads
+        attention_mask (Optional[Tensor]): Tensor of shape [b, h, d1, ..., q_dn, k_dn].
+                                           Contains 1s for positions to attend to and 0s for masked positions.
+                                           Applied before softmax.
+        head_mask (Optional[Tensor]): Tensor of shape [b, h, d1, ..., q_dn, k_dn].
```
[the remark this answers is not in the document]
Comment (RdoubleA) [reply torchmultimodal/modules/layers/attention.py:27]: this is intentional, the last dimension is q's dn and then k's dn as a result of their matrix multiplication. this is usually both the same, but I'm distinguishing here in cases like cross attention where q and k might be different dimensions
